lib/task_helpers/imports/tags.rb

```diff
@@ -0,0 +1,93 @@
+module TaskHelpers
+  class Imports
+    class Tags
+      class ClassificationDescError < StandardError; end
+      class ClassificationNameError < StandardError; end
+      class ClassificationEntryDescError < StandardError; end
+      class ClassificationEntryNameError < StandardError; end
+      class ClassificationYamlError < StandardError
+        attr_accessor :details
+
+        def initialize(message = nil, details = nil)
+          super(message)
+          self.details = details
+        end
+      end
+
+      def import(options = {})
+        return unless options[:source]
+
+        glob = File.file?(options[:source]) ? options[:source] : "#{options[:source]}/*.yaml"
+        Dir.glob(glob) do |fname|
+          begin
+            tag_categories = YAML.load_file(fname)
+            import_tags(tag_categories)
+          rescue ClassificationDescError
+            warn("Error importing #{fname} : Tag category description is required")
+          rescue ClassificationNameError
+            warn("Error importing #{fname} : Tag category name is required")
+          rescue ClassificationEntryDescError
+            warn("Error importing #{fname} : Tag entry description is required")
+          rescue ClassificationEntryNameError
+            warn("Error importing #{fname} : Tag entry name is required")
+          rescue ClassificationYamlError => e
+            warn("Error importing #{fname} : #{e.message}")
+            e.details.each { |k, v| warn("#{k}: #{v.first[:error]}") }
+          rescue ActiveModel::UnknownAttributeError => e
+            warn("Error importing #{fname} : #{e.message}")
+          end
+        end
+      end
+
+      private
+
+      # Description attribute of Tag Categories that are not visible in the UI
+      SPECIAL_TAGS = ['Parent Folder Path (VMs & Templates)', 'Parent Folder Path (Hosts & Clusters)', 'User roles'].freeze
+
+      UPDATE_FIELDS = %w(description example_text show perf_by_tag).freeze
+
+      def import_tags(tag_categories)
+        tag_categories.each do |tag_category|
+          next if SPECIAL_TAGS.include?(tag_category['description'])
+          Classification.transaction do
+            import_classification(tag_category)
+          end
+        end
+      end
+
+      def import_classification(tag_category)
+        raise ClassificationDescError unless tag_category['description']
+        raise ClassificationNameError unless tag_category['name']
```

Comment:
I don't think we need to raise exceptions here.  We already have [validations](https://github.com/ManageIQ/manageiq/blob/ca25c1575905134e930be82959c7785560c9708f/app/models/classification.rb#L12-L29) on classification.  I believe we should be able to do 

```
c = Classification.new(...) # passing the values we want to import from each classification
unless c.valid?
  # inspect c.errors to see what failed validation
end
```

I don't know if we're testing more or less things here vs. what's in the existing validations or if there's a reason we're skipping the activerecord validations and doing them manually here.

Reply:
Of course, you might have to treat "categories" different than "entries" but because they're all classifications in the end, they all must pass the existing validations so we should only have to add any additional validations you care about.  In other words, we shouldn't need to check name and description above, it should invalid and activerecord will have why it's invalid.

Reply:
These exceptions are here to prevent a user from trying to import a file that they have modified and removed the name or description. The

`classification = Classification.find_by_name(tag_category['name'])`

will generate a TypeError if `tag_category['name']` is `nil`. Having these checks here I think simplifies the error handling later in the script. The import will never succeed if either of these is `nil` so why even try. There is a check if errors were generated

`raise ClassificationYamlError.new("Tag Category error", classification.errors.details) if classification.errors.count.positive?`

to handle situations where the user changed a value to something illegal (like having a capital letter in the name).

I don't mind changing the error handling to using the `c.valid?` if that would be a better way to handle the error checking.

(In a quick test I found I can get around the `TypeError` in the `Classification.find_by_name` by using `to_s` on the `tag_category['name']` so maybe that would be better and then just handle the error of the classification not being created because the name is `nil`)

Reply:
You can do something like this if the new classification object is not valid -
```
warn("Error importing #{c.errors.full_messages.inspect") unless c.valid?
```
That'll output all the invalid attributes at one time which would eliminate the need for a user to fix them one by one. This way you're allowing the validations in the model to handle it. See if that works for you and then you can get rid of all of the special exception classes you created.

Reply:
I'm going to leave the check for `tag_category['name']` being `nil` because creating a `Classification` with `name` being `nil` causes an explosion. Having `description` `nil` or `name` being an invalid string (e.g. `inValid`) is handled by the create and the error is retuned as expected so I can get rid of those error classes.

```
irb(main):085:0> a
=> {"description"=>nil, "icon"=>nil, "read_only"=>false, "syntax"=>"string", "single_value"=>true, "example_text"=>"Brant Test", "parent_id"=>0, "show"=>true, "default"=>nil, "perf_by_tag"=>false, "name"=>nil}
irb(main):086:0> classification = Classification.create(a)
NoMethodError: undefined method `name' for nil:NilClass
	from /home/bevans/my-repos/manageiq/app/models/classification.rb:326:in `name'
	from /home/bevans/.gem/ruby/2.3.5/gems/activemodel-5.0.6/lib/active_model/validator.rb:149:in `block in validate'
	from /home/bevans/.gem/ruby/2.3.5/gems/activemodel-5.0.6/lib/active_model/validator.rb:148:in `each'
	from /home/bevans/.gem/ruby/2.3.5/gems/activemodel-5.0.6/lib/active_model/validator.rb:148:in `validate'
	from /home/bevans/.gem/ruby/2.3.5/gems/activesupport-5.0.6/lib/active_support/callbacks.rb:405:in `public_send'
	from /home/bevans/.gem/ruby/2.3.5/gems/activesupport-5.0.6/lib/active_support/callbacks.rb:405:in `block in make_lambda'
	from /home/bevans/.gem/ruby/2.3.5/gems/activesupport-5.0.6/lib/active_support/callbacks.rb:169:in `block (2 levels) in halting'
	from /home/bevans/.gem/ruby/2.3.5/gems/activesupport-5.0.6/lib/active_support/callbacks.rb:547:in `block (2 levels) in default_terminator'
	from /home/bevans/.gem/ruby/2.3.5/gems/activesupport-5.0.6/lib/active_support/callbacks.rb:546:in `catch'
	from /home/bevans/.gem/ruby/2.3.5/gems/activesupport-5.0.6/lib/active_support/callbacks.rb:546:in `block in default_terminator'
	from /home/bevans/.gem/ruby/2.3.5/gems/activesupport-5.0.6/lib/active_support/callbacks.rb:170:in `block in halting'
	from /home/bevans/.gem/ruby/2.3.5/gems/activesupport-5.0.6/lib/active_support/callbacks.rb:454:in `block in call'
	from /home/bevans/.gem/ruby/2.3.5/gems/activesupport-5.0.6/lib/active_support/callbacks.rb:454:in `each'
	from /home/bevans/.gem/ruby/2.3.5/gems/activesupport-5.0.6/lib/active_support/callbacks.rb:454:in `call'
	from /home/bevans/.gem/ruby/2.3.5/gems/activesupport-5.0.6/lib/active_support/callbacks.rb:101:in `__run_callbacks__'
	from /home/bevans/.gem/ruby/2.3.5/gems/activesupport-5.0.6/lib/active_support/callbacks.rb:750:in `_run_validate_callbacks'
... 17 levels...
	from /home/bevans/.gem/ruby/2.3.5/gems/activerecord-5.0.6/lib/active_record/connection_adapters/abstract/database_statements.rb:232:in `transaction'
	from /home/bevans/.gem/ruby/2.3.5/gems/activerecord-5.0.6/lib/active_record/transactions.rb:211:in `transaction'
	from /home/bevans/.gem/ruby/2.3.5/gems/activerecord-5.0.6/lib/active_record/transactions.rb:392:in `with_transaction_returning_status'
	from /home/bevans/.gem/ruby/2.3.5/gems/activerecord-5.0.6/lib/active_record/transactions.rb:319:in `block in save'
	from /home/bevans/.gem/ruby/2.3.5/gems/activerecord-5.0.6/lib/active_record/transactions.rb:334:in `rollback_active_record_state!'
	from /home/bevans/.gem/ruby/2.3.5/gems/activerecord-5.0.6/lib/active_record/transactions.rb:318:in `save'
	from /home/bevans/.gem/ruby/2.3.5/gems/activerecord-5.0.6/lib/active_record/suppressor.rb:41:in `save'
	from /home/bevans/.gem/ruby/2.3.5/gems/activerecord-5.0.6/lib/active_record/persistence.rb:34:in `create'
	from (irb):86
	from /home/bevans/.gem/ruby/2.3.5/gems/railties-5.0.6/lib/rails/commands/console.rb:65:in `start'
	from /home/bevans/.gem/ruby/2.3.5/gems/railties-5.0.6/lib/rails/commands/console_helper.rb:9:in `start'
	from /home/bevans/.gem/ruby/2.3.5/gems/railties-5.0.6/lib/rails/commands/commands_tasks.rb:78:in `console'
	from /home/bevans/.gem/ruby/2.3.5/gems/railties-5.0.6/lib/rails/commands/commands_tasks.rb:49:in `run_command!'
	from /home/bevans/.gem/ruby/2.3.5/gems/railties-5.0.6/lib/rails/commands.rb:18:in `<top (required)>'
	from bin/rails:4:in `require'
	from bin/rails:4:in `<main>'
```

Reply:
I've found a better way to handle a `name` being `nil` during import. I'm going to do a `to_s` before the import for the `name`.